aip/0162.md

```diff
@@ -0,0 +1,229 @@
+---
+aip:
+  id: 162
+  state: draft
+  created: 2019-09-17
+permalink: /162
+redirect_from:
+  - /0162
+---
+
+# Resource Revisions
+
+Some APIs need to have resources with a revision history, where users can
+reason about the state of the resource over time. There are several reasons for
+this:
+
+- Users may want to be able to roll back to a previous revision, or diff
+  against a previous revision.
+- An API may create data which is derived in some way from a resource at a
+  given point in time. In these cases, it may be desirable to snapshot the
+  resource for reference later.
+
+## Guidance
+
+APIs **may** store a revision history for a resource if it is useful to users.
+
+APIs implementing resources with a revision history **must** provide a
+`revision_id` field on the resource:
+
+```proto
+message Book {
+  // The name of the book.
+  string name = 1;
+
+  // Other fields…
+
+  // The revision ID of the book.
+  // This is automatically changed whenever the contents of the book
+  // change in any way.
+  string revision_id = 99 [(google.api.field_behavior) = OUTPUT_ONLY];
+}
+```
+
+- The `revision_id` field **should** be a string and contain a short,
```

Comment:
There are two alternatives:

- An auto-incrementing counter (1, 2, 3...)
- Dates

Ultimately, I think an unordered hash string is probably the wisest course, but that is definitely up for debate. Having worked with revision control systems that use both methods, hashes provide the most flexibility, especially if an API has a situation where history needs to be mutable (e.g. delete a revision that is no longer needed), and that has come up in several cases.

Reply:
Yeah, the intent is that you get the chronologically-ordered list with `ListBookRevisions`.

Reply:
The AutoML team adds a third alternative: semantic versioning.

I find this alternative silly -- unless the service can determine compatibility between resource B and resource A (which seems vanishingly unlikely), semver is meaningless. This feels like a footgun to me.

Reply:
In general we would like to propose allowing API writers to decide the format of the versions, there are many version schemes in the wild, and at least some of them have reasonable use-cases.

E.g. in AutoML for versioning Models, we would probably let users  pick if they want versions of given Model to be generated by our infra (probably default), or should it be a string that the user has to populate each time themselves. This way user's that need semantic versioning could effectively represent it. Otherwise, the users would use some labels on Model to do the same thing, but I think that's uglier and annoying to the user.

Reply:
Discussed in-person:
@jgeewax and I want to make sure we do not boil the ocean, and we are not confident that this use case applies outside AutoML. We think it is probably fine for AutoML to do it but we do not want to specify it as a general pattern. @rsurowka has agreed with this in principle.

Reply:
@jgeewax and I agree that git-style hashes are preferable over auto-incermenting IDs or dates.